libcodechecker/server/client_db_access_handler.py

```diff
@@ -93,13 +105,126 @@ def bugreportpoint_db_to_api(brp):
+        return DetectionStatus.REOPENED
+
+
+def unzip(b64zip):
+    """
+    This function unzips the base64 encoded zip file which should contain
+    exactly one directory. This directory is extracted to the given workspace.
+    The function returns the name of the extracted directory.
+    """
+
+    _, zip_file = tempfile.mkstemp('.zip')
+
+    with open(zip_file, 'wb') as zip_f:
+        zip_f.write(base64.b64decode(b64zip))
+
+    zipf = zipfile.ZipFile(zip_file, 'r')
+
+    # A not too nice way to find the only directory in the zip archive.
+    first_file = zipf.namelist()[0]
+    zip_dir = first_file[:first_file.find('/')]
+
+    zipf.extractall('/tmp')
+    zipf.close()
+
+    os.remove(zip_file)
+
+    return os.path.join('/tmp', zip_dir)
+
+
+class StorageSession:
+    """
+    This class is a singleton which helps to handle a transaction which
+    belong to the checking of an entire run. This class holds the SQLAlchemy
+    session for the run being checked and the set of touched reports. This
+    latter one is needed so at the end the detection status of the rest reports
+    can be set to "resolved".
+    """
+
+    class __StorageSession:
+        def __init__(self):
+            self.__sessions = dict()
+            self._timeout_sessions()
+
+        def start_run_session(self, run_id, transaction):
+            self.__sessions[run_id] = {
+                'touched_reports': set(),
+                'transaction': transaction,
+                'timer': time.time()}
+
+        def end_run_session(self, run_id):
+            this_session = self.__sessions[run_id]
+            transaction = this_session['transaction']
+
+            # Set resolved reports
+
+            transaction.query(Report) \
+                .filter(Report.run_id == run_id,
+                        Report.id.notin_(this_session['touched_reports'])) \
+                .update({Report.detection_status: 'resolved'},
+                        synchronize_session='fetch')
+
+            transaction.commit()
+            transaction.close()
+
+            del self.__sessions[run_id]
+
+        def abort_session(self, run_id):
+            transaction = self.__sessions[run_id]['transaction']
+            transaction.rollback()
+            transaction.close()
+            del self.__sessions[run_id]
+
+        def touch_report(self, run_id, report_id):
+            self.__sessions[run_id]['touched_reports'].add(report_id)
+
+        def is_touched(self, run_id, report_id):
+            return report_id in self.__sessions[run_id]['touched_reports']
+
+        def has_ongoing_run(self, run_id):
+            return run_id in self.__sessions
+
+        def get_transaction(self, run_id):
+            self.__sessions[run_id]['timer'] = time.time()
+            return self.__sessions[run_id]['transaction']
+
+        def _timeout_sessions(self):
+            for run_id, session in self.__sessions.iteritems():
+                if int(time.time() - session['timer']) > 10:
```

Comment:
Does this mean that 10 seconds after the last call, e.g. `addReport()` calls on the session, it times out? I think this is worthy of an explaining comment here.